### PR TITLE
[WIP] Adding extra detail to error messages

### PR DESCRIPTION
**Pull request by cloudRoutine**
Implements more detailed error messages using formatted strings that contain the current error message and specifics about why the exception was thrown.

Described in detail - https://github.com/Microsoft/visualfsharp/issues/1438

kevinr -- Link with whitespace changes ignored: https://github.com/Microsoft/visualfsharp/pull/1454/files?w=1

### CODE REVIEW

**Comment by cloudRoutine** [comment]
@Microsoft/fsharp-compiler I think this is at the point where it could use some review.

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Collections/CollectionModulesConsistency.fs [hunk 1]
```diff
@@ -8,11 +8,17 @@ open NUnit.Framework
 open FsCheck
 open Utils
 
+/// helper function that creates labeled FsCheck properties for equality comparisons
+let consistency name sqs ls arr =
```
Comment (smoothdeveloper) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Collections/CollectionModulesConsistency.fs]: this is good idea to replace all the assertions with duplicated logic and have better error message.

Should it be `checkConsistency` or even `assertConsistency` (I assume it raises an exception if the equality is not observed).

File: src/fsharp/FSharp.Core/array2.fs [hunk 2]
```diff
@@ -130,17 +131,27 @@ namespace Microsoft.FSharp.Collections
             init (length1 array) (length2 array) (fun i j -> array.[b1+i,b2+j])
 
         [<CompiledName("CopyTo")>]
-        let blit (source : 'T[,])  sourceIndex1 sourceIndex2 (target : 'T[,]) targetIndex1 targetIndex2 count1 count2 = 
+        let blit (source : 'T[,])  sourceIndex1 sourceIndex2 (target: 'T[,]) targetIndex1 targetIndex2 count1 count2 =
```
Comment (smoothdeveloper) [inline src/fsharp/FSharp.Core/array2.fs]: I gather that error message will now report bound per dimensions when there is a mismatch, should tests be added to assert this behaviour does what is intended wrt. new error messages?

**Comment by smoothdeveloper** [comment]
I like those changes to get more user-friendly and useful error message in the core library and I like the pass of slight refactoring to reduce duplication / better encapsulation of the logic building those messages.

👍

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Collections/CollectionModulesConsistency.fs (hunk 1 above)
Comment (cloudRoutine) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Collections/CollectionModulesConsistency.fs]: by default it does not, if run with `Check.QuickThrowOnFailure` that method will throw the exception when the `Property` returns a failing case.

**Comment by cloudRoutine** [comment]
@dotnet-bot test this please

**Comment by cloudRoutine** [comment]
@KevinRansom  I tried to set this up to make it easier for localization, but I'm not sure how to take this forward.

The only places where any descriptive text is used apart from the names of the function's arguments are in [the DetailedExceptions module](https://github.com/cloudRoutine/visualfsharp/blob/improve-errors/src/fsharp/FSharp.Core/local.fs#L20-L71) and [the axis & bound names for Array2D](https://github.com/cloudRoutine/visualfsharp/blob/improve-errors/src/fsharp/FSharp.Core/array2.fs#L133-L158)

Is there anything else I need to do to get this PR merge-ready?

File: src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Collections/CollectionModulesConsistency.fs [hunk 3]
```diff
@@ -8,11 +8,17 @@ open NUnit.Framework
...
+    (sqs = arr) |@ (sprintf  "Seq.%s = %A, Array.%s = %A" name sqs name arr) .&.
```
Comment (KevinRansom) [inline src/fsharp/FSharp.Core.Unittests/FSharp.Core/Microsoft.FSharp.Collections/CollectionModulesConsistency.fs]: can you put ticks around the %A I.e. '%A'  it allows us to see that a value is missing in the event of bugs.

**Comment by KevinRansom** [comment]
@cloudRoutine localization is not needed for fieldnames, or argument names.

File: src/fsharp/FSharp.Core/local.fs [hunk 4]
```diff
@@ -1,8 +1,80 @@
 // Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+
+namespace Microsoft.FSharp.Core
+
+module internal DetailedExceptions =
```
Comment (KevinRansom) [inline src/fsharp/FSharp.Core/local.fs]: Perhaps [<AutoOpen>] to eliminate all of the opens?  It seems like these are generally applicable.

File: src/fsharp/FSharp.Core/local.fsi [hunk 5]
```diff
@@ -1,5 +1,21 @@
 // Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+namespace Microsoft.FSharp.Core
+open Microsoft.FSharp.Core
+
+module internal DetailedExceptions =
```
Comment (KevinRansom) [inline src/fsharp/FSharp.Core/local.fsi]: AutoOpen?

File: src/fsharp/FSharp.Core/seq.fs [hunk 6]
```diff
@@ -360,145 +364,145 @@ namespace Microsoft.FSharp.Collections
...
                 match g with
                 | :? GenerateThen<'T> as g -> GenerateThen<_>.Bind(g.Generator,(fun () -> GenerateThen<_>.Bind (g.Cont(), cont)))
                 | g -> (new GenerateThen<'T>(g, cont) :> Generator<'T>)
 
 
         let bindG g cont = GenerateThen<_>.Bind(g,cont)
 
-        //let emptyG () = 
-        //    { new Generator<_> with 
+        //let emptyG () =
```
Comment (KevinRansom) [inline src/fsharp/FSharp.Core/seq.fs]: I think we should just get rid of the commented out code.

**Comment by KevinRansom** [comment]
The changes look good, I left a few comments to consider.

Both @forki and @dsyme complain whenever I do these opportunistic cleanups within a PR that does a specific change.  Note:  I personally don't mind it honest.

When the opportunistic cleanup is extensive like it is in this PR they both submit them as separate cleanup PRs.  Don't change this one ... but next time if you see a cleanup harvest just submit it as a separate cleanup only PR.

Thanks for taking care of this

Kevin

**Comment by cloudRoutine** [comment]
I'll keep the cleanup in mind for future work, thanks for the review

**Comment by cloudRoutine** [comment]
btw I made all the changes you suggested

**Comment by KevinRansom** [comment]
@cloudRoutine 
Thank you for taking care of this.

Kevin

**Pull request merged**